21.md

```diff
@@ -0,0 +1,210 @@
+- Omitting the `source` prevents the receiver from knowing the cost of the BTC being sent
+- Omitting the `target` implies the receiver will get BTC from the payment, no matter the `source` or `amount` denomination
+
+Note that the amount provided in all requests is always an integer number interpreted as the smallest unit of the selected `currency`. The smallest unit needs to be according to the `decimals` parameter, so the `WALLET` has all the needed information to receive input and show output properly.
+
+### Service-side second response
+
+Upon receiving a currency-denominated request from `WALLET`, the `SERVICE` must return an invoice with an amount matching the converted rate for the amount in that currency. The rate used does not need to match the `multiplier` first informed.
+
+If the `WALLET` requested an actual conversion, the `SERVICE` must provide an additional field alongside the invoice informing the guaranteed `converted` amount that will be creditted to the receiver upon payment. The `converted` amount, and therefore the conversion rate, must be guaranteed by the `SERVICE` for as long as the invoice is not expired. The `converted` amount must be denominated in the smallest unit of the currency, just like the `amount` parameter.
+
+If the `WALLET` informed the `SERVICE` about the cost of the BTC used in the payment, the `SERVICE` may use the information to display currency-to-currency exchange rates.
+
+```typescript
+type BaseResponse = {
+  pr: string,
+  routes: [],
+}
+
+type ExtendedResponse = BaseResponse & {
+  converted?: number, // Integer; Present if and only if `target` was received.
+}
+```
+
+```diff
+{
+  "pr": "lnbc1230n1pjknkl...ju36m3lyytlwv42fee8gpt6vd2v",
+  "routes": [],
++ "converted": 123
+}
+```
+
+### Examples
+These examples show all the possible uses of this extension by a supporting `WALLET` and `SERVICE`.
+
+#### Payer queries the payee service
+`GET <service>/.well-known/lnurlp/<identifier>`
+```json
+{
+  "tag": "payRequest",
+  "callback": "bipa.app/callback",
+  "metadata": "...",
+  "minSendable": 1000,
+  "maxSendable": 1000000,
+  "currencies": [
+    {
+      "code": "BRL",
+      "name": "Reais",
+      "symbol": "R$",
+      "decimals": 2,
+      "multiplier": 5405.405,
+      "convertible": true
+    },
+    {
+      "code": "USDT",
+      "name": "Tether",
+      "symbol": "₮",
+      "decimals": 6,
+      "multiplier": 26315.789
+    }
+  ]
+}
+```
+###### Payer sends 538 sats
+```json5
+// GET <callback>?amount=538000
+{ "pr": "(invoice of 538 sats)" }
+```
+###### Payer sends 1 BRL worth of BTC
+```json5
+// GET <callback>?amount=100.BRL
+{ "pr": "(invoice of 538 sats)" }
+```
+###### Payer sends 538 sats to be converted into BRL
+```json5
+// GET <callback>?amount=538000&target=BRL
+{ "converted": 100, "pr": "(invoice of 538 sats)" }
+```
+###### Payer sends 538 sats that cost 0.2 USDT
+```json5
+// GET <callback>?amount=538000&source=200000.USDT
+{ "pr": "(invoice of 538 sats)" }
```

Comment:
One thing I might be missing here - in what cases would the SERVICE care what the source cost+currency was? How is that information used? Is there some legal reason this is needed for remittences or something?

Reply:
The SERVICE might care in order to provide a better experience, in this case, more information about the operation. For example, if you are service A sending dollars to service B to be converted into euros.

The sending service (A) is gonna buy the dollars and the user will confirm the quote for selling into euros, so A will know the whole exchange, including the dollar-to-euro conversion rate.

The receiving service (B) of course knows about the sell, but it wouldn't know about the dollars without being informed by A. Having this information exchanged means both sides can say the actual dollar-to-euro exchange rate and show the information to the user.

Two random examples I can think of are, top of my head: you're actually sending money to yourself overseas or something like transferwise and you'd want to have information about the exchange on both sides. Second one would be you're paying someone an agreed upon amount in your local currency, like I need to pay you $5 for this comment. Trust both services implementations, your wallet would know that I paid $5 for the BTC I sent you, no matter how many sats you got on your side.

Does that explain the logic?

Reply:
That's exactly what I had in mind. Exactly that!